### PR TITLE
Hash SCIM tokens

### PR DESCRIPTION
**Pull request by smatting**
Currently SCIM tokens are persisted in plain text. Fixes https://github.com/zinfra/backend-issues/issues/1652

With this PR new SCIM tokens will only have their SHA512 hashes persisted.
Existing tokens (still persisted in plaintext) will remain valid, but will be re-persisted with their hash on every validation.

The columns `team_provisioning_by_token.token_`, `team_provisioning_by_team.token_` which contained the plaintext token will be repurposed and also contain the hash value.
To distinguish plaintext from hash values the hash values are prefixed with `sha512:`. In Haskell these two types are represented as the sum type

```Haskell
data ScimTokenLookupKey
  = ScimTokenLookupKeyHashed ScimTokenHash
  | ScimTokenLookupKeyPlaintext ScimToken
  deriving (Eq, Show)
```

This way all plaintext tokens which are regularly used should be converted to their hash. The remaining tokens should be converted via batch process or invalidated.

- [ ] create follow-up issue for that

### CODE REVIEW

File: services/spar/src/Spar/Types.hs [hunk 1]
```diff
@@ -143,6 +147,25 @@ instance ToJSON IdPMetadataInfo where
...
   deriving (Eq, Show, FromJSON, ToJSON, FromByteString, ToByteString)
 
+newtype ScimTokenHash = ScimTokenHash {fromScimTokenHash :: Text}
+  deriving (Eq, Show)
+
+instance FromByteString ScimTokenHash where
+  parser = string "sha512:" *> (ScimTokenHash <$> parser)
+
+instance ToByteString ScimTokenHash where
+  builder (ScimTokenHash t) = BB.fromByteString "sha512:" <> builder t
```
Comment (fisx) [inline services/spar/src/Spar/Types.hs:157]: Add a roundtrip test, like [these](https://github.com/wireapp/wire-server/blob/3af1aa63a9cb4d6594a4e017b0ba38408c0f26d2/libs/wire-api/test/unit/Test/Wire/API/Roundtrip/ByteString.hs)?

(I admit that there are more gaps in the roundtrip test coverage, but we might as well start now to patch them...)

File: services/spar/src/Spar/Data.hs [hunk 2]
```diff
@@ -619,15 +623,36 @@ lookupScimToken ::
...
+  ScimTokenInfo ->
+  m ()
+connvertPlaintextToken token ScimTokenInfo {..} = retry x5 . batch $ do
+  setType BatchLogged
+  setConsistency Quorum
+  addPrepQuery delById (stiTeam, stiId)
+  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
+  let tokenHash = hashScimToken token
+  addPrepQuery insByToken (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
+  addPrepQuery insByTeam (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
```
Comment (fisx) [inline services/spar/src/Spar/Data.hs:655]: Deletion of the old entries must happen after insertion of the new ones, to handle crashes in between.  Both, or even either of the rows being hashed while there is still an unhashed
```suggestion
  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
  let tokenHash = hashScimToken token
  -- enter by new lookup key
  addPrepQuery insByToken (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
  -- update info table
  addPrepQuery insByTeam (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
  -- remove old lookup key
  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
```

File: services/spar/src/Spar/Data.hs [hunk 3]
```diff
@@ -619,15 +623,36 @@ lookupScimToken ::
   ScimToken ->
   m (Maybe ScimTokenInfo)
 lookupScimToken token = do
-  mbRow <- retry x1 . query1 sel $ params Quorum (Identity token)
-  pure $ fmap fromScimTokenRow mbRow
-  where
-    sel :: PrepQuery R (Identity ScimToken) ScimTokenRow
+  let tokenHash = hashScimToken token
+  mbRow <- retry x1 . query1 sel $ params Quorum (tokenHash, token)
```
Comment (fisx) [inline services/spar/src/Spar/Data.hs:627]: This can return both hashed and unhashed in any order, if both are still in the DB (race condition).

**Review by fisx — COMMENTED** [review]
:+1:  (some comments)

File: services/spar/src/Spar/Types.hs (hunk 1 above)
Comment (fisx) [reply services/spar/src/Spar/Types.hs:157]: (we can also do that in a separate PR)

Comment (smatting) [reply services/spar/src/Spar/Types.hs:157]: added roundtrip test

File: services/spar/src/Spar/Data.hs (hunk 3 above)
Comment (smatting) [reply services/spar/src/Spar/Data.hs]: done

File: services/spar/src/Spar/Data.hs [hunk 4]
```diff
@@ -619,15 +625,48 @@ lookupScimToken ::
...
+
+connvertPlaintextToken ::
+  (HasCallStack, MonadClient m) =>
+  ScimToken ->
+  ScimTokenInfo ->
+  m ()
+connvertPlaintextToken token ScimTokenInfo {..} = retry x5 . batch $ do
+  setType BatchLogged
+  setConsistency Quorum
+  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
```
Comment (fisx) [inline services/spar/src/Spar/Data.hs:662]:
```suggestion
```

This looks like it was a left-over?

**Review by fisx — APPROVED** [review]
Found two things (both harmless).  Integration tests pass locally.  :shipit:

**Comment by fisx** [comment]
> Found two things (both harmless). Integration tests pass locally. :shipit:

That was wrong: the stray delete query risks a race condition where a token that should be available is already gone in unhashed from and not there yet in hashed form.  Not very likely to materialize, but please fix it anyway.

**Comment by smatting** [comment]
`1 out of 247 tests failed (356.63s)`
The failing test is unrelated to this PR: https://github.com/zinfra/backend-issues/issues/1875

**Pull request merged**